---
title: "Homework 1"
author: "your name here"
date: "`r format(Sys.time(), '%m/%d/%Y')`"
output: html_document
---

# Question 1 - 8 pts

In week 2 (EDA), remember that we looked at RNA from different sub-regions of
the brain, from the GTEx Project. This week we learned that, the first time we
looked at the data, we did it the "wrong" way, in that we broke the SE object
apart into `condition` (sample labels) and `cts` (a matrix of observations of
RNA transcripts, of dimension genes x samples).

In this question you should redo that analysis, but this time, don't break apart
the SE object, but keep all the information in the SE object. That is, instead
of using `condition` (a stand alone vector) you should work with `se$condition`,
a column of `colData(se)`. This helps to prevent mistakes where you subset
either the matrix or the sample labels, but forget to perform the same operation
on the other, as the SE keeps all the information tied together.

First load the data from the brain RNA lecture note (you may need to change the
`source()` line, just copy the `my_scale_counts.R` scripts from `bioc` directory
into your homework repo).

```{r}
library(here)
url <- "http://duffel.rail.bio/recount/SRP012682/rse_gene_brain.Rdata"
file <- here("bioc","brain.rda")
if (!file.exists(file)) download.file(url, file)
load(file)
suppressPackageStartupMessages(library(SummarizedExperiment))
source(here("bioc","my_scale_counts.R"))
se <- my_scale_counts(rse_gene)
```

Then create the condition variable as a metadata column:

```{r}
se$condition <- sub("Brain - (.*)", "\\1", se$smtsd)
table(se$condition)
```

Subset the SE to these four regions below, then make a PCA plot of log2(counts +
1) for the four regions. 

* The PCA plot should be colored by region
* The PCA plot should just look at the top 500 genes by variance

```{r}
regions <- c("Cerebellum","Cortex","Hippocampus","Hypothalamus")
```

Hint: one simple way to do this is to replace the `assay()` with log counts,
pass the SE to `DESeqTransform` and then use `DESeq2::plotPCA` to make a plot.
Or you can do this manually if you like. Note that `plotPCA` automatically
subsets the data to the top 500 genes by variance.

Are there any samples that may indicate sample swap / mis-labelling?

# Question 2 - Gene density in 1 Mb tiles - 8 pts

We want to find out if genes are roughly uniform or follow a pattern
along the genome. E.g. do they occur in clumps?

Start by loading the Ensembl database package we used in class 
`EnsDb.Hsapiens.v86`. For simplicity, rename this to `edb`.

```{r message=FALSE}
library(EnsDb.Hsapiens.v86)
edb <- EnsDb.Hsapiens.v86
```

Pull out the genes as a simple *GRanges* object. We will use a special
command which will restrict our focus to genes with IDs that start
with `ENSG`:

```{r}
g <- genes(edb, 
  filter = AnnotationFilterList(GeneIdFilter("ENSG", "startsWith")))
```

Subset to the genes on chromosome `"1"` (there should be 5194 of these). 
Next you will make a `query` GRanges which covers
chromosome 1 with non-overlapping ranges of width 1 Mb (one million
letters). This can be done with the following code:

```{r}
query <- tileGenome(seqlengths(g)[1], 
                    tilewidth=1e6, 
                    cut.last.tile.in.chrom=TRUE)
```

Count the number of genes overlapping each range (you can use either
`countOverlaps` or `plyranges::count_overlaps`) 
and then plot the counts (after converting
to a numeric vector). You can just ignore the fact that the last tile
is shorter than the rest.

Do you see tile-to-tile correlations in the counts along chromosome 1 
or do the counts look independent from tile-to-tile, with no
sequential dependency? 

What is the average number of genes per 1 Mb (you can just include the
last tile count in the calculation even though it is not 1 Mb)?

Again, compute the average number of genes but exclude the tiles with
zero genes (some of these are telomeric or centromeric DNA).

