---
title: "Homework 2"
author: "your name here"
date: "`r format(Sys.time(), '%m/%d/%Y')`"
output: html_document
---

# Question 1 - Gene density and GC content - 8 pts

Does gene density as calculated in Question 1 relate to the 
"GC content", defined as the ratio of C and G
bases in a region? Remember, GC content = sum(C or G bases) divided
by the sequence length. GC content of ACGT = 0.5, GC content of ACGC =
0.75.

Assess the GC content of the tiled chromosome 1
with `query` as defined above, and assess whether it appears related
to gene density. How would you describe the relationship?

Use the following BSgenome package:

```{r}
suppressPackageStartupMessages(library(BSgenome.Hsapiens.NCBI.GRCh38))
suppressPackageStartupMessages(library(Biostrings))
```

# Question 2 - Exon and gene GC content by gene type - 8 pts

In this question we will examine the exonic GC content and the 
gene-level GC content, per gene. The exonic GC content is the GC content
of any basepair that is in any exon of the gene. The gene GC content
will be the GC content of the entire sequence including exons and introns.

Start by loading the exons per gene, subsetting to standard chromosomes,
and then `reduce`-ing the exon-by-gene GRangesList object:

```{r}
ebg <- exonsBy(edb, by="gene",
               filter = 
             AnnotationFilterList(GeneIdFilter("ENSG", "startsWith")))
ebg <- keepStandardChromosomes(ebg, pruning.mode = "coarse")
ebg <- reduce(ebg)
```

What does `reduce` do? It basically "flattens" the ranges so that each
basepair is only represented in one range. For example compare `foo`
to `reduce(foo)`: 

```{r}
foo <- IRanges(start=c(1,5,11),end=c(7,9,13))
foo
reduce(foo)
```

So calling `reduce` on the exons for each gene "flattens" the ranges
so that any overlapping basepairs are compressed into a single
continguous range. 

Now repeat the same subset for the genes (as defined above), and make
sure the two are in the same order.

```{r}
g <- keepStandardChromosomes(g, pruning.mode = "coarse")
g <- g[names(ebg)]
all.equal(names(g), names(ebg))
```

We can use the functions we learned in class to get the DNA
sequence of genes and exons:

```{r}
library(GenomicFeatures)
gseqs <- getSeq(Hsapiens, g)
eseqs <- extractTranscriptSeqs(Hsapiens, ebg)
```

Note that you can get the number of basepairs of the exonic region
per gene two ways:

```{r}
all(sum(width(ebg)) == width(eseqs))
```

Compare the exonic GC content across two types of genes:
`protein_coding`, and `lincRNA` (see the metadata of `g`). 
Repeat for the gene GC content.
It is sufficient to compare visually.

